### PR TITLE
MeshUVSpaceRenderer UV Edge Blending 

### PR DESCRIPTION
**Pull request by torchesburn**
This PR fixes the UV island seam issue by introducing an optional blending step where the edges of the UVs are blended, enabled using the flag: uvEdgeBlending. Very useful for UV islands.

image

image

## How to enable the Flag:
mesh.decalMap = new MeshUVSpaceRenderer(m, scene, {width: 4096, height: 4096, uvEdgeBlending: true});

## Caveats
RTT must have same dimensions as the Mesh UV Space Renderer for the UV Edge Blending to work.
There may be a way around this, but it would make sense to let the user know through JSDoc comment.

## Optimisations
Mask texture, and final post processing effect only generated once

## Playground to test:
http://localhost:1338/#N10DXG#365

The following guidelines have been followed:
https://github.com/BabylonJS/Babylon.js/blob/d01337f9d3428033e4d723f9a10187bfddf381e5/contributing.md

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by Popov72** [comment]
Here are some general comments you should address before I review more closely:
* don't use "any" as types
* use the `public`/`private` access modifier as appropriate (some properties don't have any)
* don't make `renderTexture` async, it is not needed
* don't create a new mask material per `meshUVSpaceRenderer` instance. You should do much like `_GetShader` (something like `_GetMaskShader`)
* remove the `finalMaterial`, which is not needed
* remove the try/catch constructs, as it's not expected there could be any error in this code (and if there are, we prefer they are issued directly from the faulty code)
* make sure that there's no breaking change with the `texture` property. Currently, the user can set it to use his own texture instead of letting the class creates one. In the current PR, it is not possible.

**Comment by sebavan** [comment]
marking as draft in the meantime to prevent accidental merge.

**Comment by torchesburn** [comment]
> * don't use "any" as types
> * use the `public`/`private` access modifier as appropriate (some properties don't have any)
> * don't make `renderTexture` async, it is not needed
> * don't create a new mask material per `meshUVSpaceRenderer` instance. You should do much like `_GetShader` (something like `_GetMaskShader`)
> * remove the `finalMaterial`, which is not needed
> * remove the try/catch constructs, as it's not expected there could be any error in this code (and if there are, we prefer they are issued directly from the faulty code)
> * make sure that there's no breaking change with the `texture` property. Currently, the user can set it to use his own texture instead of letting the class creates one. In the current PR, it is not possible.

I have addressed each of these points, especially the last one. 

User should be able to create their own RTT and assign it to the MeshUVSpaceRenderer, and enable uvEdgeBlending or not, or let the system internally create their texture and enable uvEdgeBlending or not.

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:218]: Can you use MeshUVSpaceRenderer._isRenderTargetTexture instead for perf reasons ?

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:240]: Should it not check _userCreatedTextureSetup instead so if the user provides an RTT it does not break its setup ?

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:245]: you can remove the following as you call addOnce, the obsevable will be "cleaned up" automatically. Adding an extra empty function is only decreasing perf.

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:317]: just use Effect in the typings here :-)

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:320]: Use setFloat2 instead to reduce GC from creating the Vector2

**Review by sebavan — CHANGES_REQUESTED** [review]
Thanks a lot for the contrib and the issues you already addressed !!!

Just a first quick pass on the PR and I ll review with test more in detail after the requested changes will be in,

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (torchesburn) [reply packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts]: I took this from the existing _createRenderTargetTexture function which was using it to generate the decal texture and from the comment next to it:

 // this disables clearing the texture for the next frames

**Comment by torchesburn** [comment]
I've resolved all conversations except for one which I've left a comment for the reason why its still in use.

Thanks for the PR feedback, the rest of the comments have been resolved.

Look forward to further testing.

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (Popov72) [reply packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts]: Yes, you should keep the empty function, it's the current way to avoid the RTT to be automatically cleared (we should probably add a property for this one day!).

**Review by sebavan — COMMENTED** [review]
LGTM lets wait on @Popov72 final approval to merge it in.

I will do today the autoclear on RTT cause yeah....

**Comment by torchesburn** [comment]
> I was reviewing the PR, and it occurred to me that we could do this more easily, by simply adding the post-process to the existing texture...
> 
> Instead of making a lot of comments, I preferred to do it locally, so here are my changes (I just took your code and rearranged it a bit - it's now less than half the changes in the original PR):
> 
> [code changes](https://github.com/BabylonJS/Babylon.js/compare/master...Popov72:Babylon.js:decaltexture-fix-seams-new)
> 
> If you agree, you can update your PR with the new changes.

That is fantastic, thanks for doing that. Will test it out

**Comment by torchesburn** [comment]
So I've updated the class with your feedback changes. 

I had to add an extra function to configure already existing user RTT, and also had to clear the post process effect.

http://localhost:1338/#N10DXG#394

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:198]: the first this.texture is not useful here as we are in the else case I believe.

**Review by sebavan — COMMENTED** [review]
LGTM !!! waiting on final approval from @Popov72 and the tiny comment

**Comment by RaananW** [comment]
@sebavan - can I reset your review? you are marked as "request changes", but have enabled auto-merge, so I am not sure which one is it :-)

**Comment by Popov72** [comment]
Please don't auto-merge, I still have to review the changes!

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (Popov72) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:257]: As far as I'm concerned, it's up to the user whether or not to clear the RTT (after all, he may want to start with an uncleared texture). You can therefore replace this code with :
```typescript
this.texture.onClearObservable.add(() => {});
```

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (Popov72) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:275]: You must also set `this._userCreatedTextureConfigured = true;`, otherwise `_configureUserCreatedRTT` will also be called when the texture is created internally (and you'll get WebGL errors in the console).

To be clearer, I'd rename `_userCreatedTextureConfigured` to `_configureUserCreatedTexture` and reverse the operation: I'd initialize it to `true` and set it to `false` in `_configureUserCreatedRTT` and `_createDiffuseRTT`.

File: packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts
Comment (Popov72) [inline packages/dev/core/src/Meshes/meshUVSpaceRenderer.ts:281]:
```suggestion
            texture.addPostProcess(this._finalPostProcess!);
```

**Review by Popov72 — CHANGES_REQUESTED** [review]
Great, almost there!

**Review by sebavan — APPROVED** [review]

**Pull request merged**
